book/src/super-sql/intro.md

```diff
@@ -1,5 +1,459 @@
+reminiscent of a `table.column` reference in SQL.
+
+For example, the SQL query from above can thus be written in pipe form
+using the [`values`](operators/values.md) operator as:
+```
+values {x:1}, {x:2}, {x:3} | select this.x
+```
+which results in:
+```
+{x:1}
+{x:2}
+{x:3}
+```
+As with SQL table names, where `this` is implied, it is optional can be omitted, i.e.,
+```
+values {x:1}, {x:2}, {x:3} | select x
+```
+produces the same result.
+
+Referencing `this` is often convenient, however, as in this query
+```
+values {x:1}, {x:2}, {x:3} | aggregate collect(this)
+```
+which collects each input value into an array and emits the array resulting in
+```
+[{x:1},{x:2},{x:3}]
+```
+
+#### Combining Piped Data
+
+If all data for all operators were always presented as a single input sequence
+called `this`, then there would be no way to combine data from different entities,
+which is otherwise a hallmark of SQL and the relational model.
+
+To remedy this, SuperSQL extends pipe scoping to
+[_joins_](#join-scoping) and
+[_subqueries_](#subquery-scoping)
+where multiple entities can be combined into the common value `this`.
+
+##### Join Scoping
+
+To combine joined entities into `this` via pipe scoping, the
+[`join`](operators/join.md) operator
+includes an _as clause_ that names the two sides of the join, e.g.,
+```
+... | join ( from ... ) as {left,right} | ...
+```
+Here, the joined values are formed into a new two-field record
+whose first field is `left` and whose second field is `right` where the
+`left` values come from the parent operator and the `right` values come
+from the parenthesized join query argument.
+
+For example, suppose the contents of a file `f1.json` is
+```
+{"x":1}
+{"x":2}
+{"x":3}
+```
+and `f2.json` is
+```
+{"y":4}
+{"y":5}
+```
+then a `join` can bring these two entities together into a common record
+which can then be subsequently operated upon, e.g.,
+```
+from f1.json
+| cross join (from f2.json) as {f1,f2}
+```
+computes a cross-product over all the two sides of the join
+and produces the following output
+```
+{f1:{x:1},f2:{y:4}}
+{f1:{x:2},f2:{y:4}}
+{f1:{x:3},f2:{y:4}}
+{f1:{x:1},f2:{y:5}}
+{f1:{x:2},f2:{y:5}}
+{f1:{x:3},f2:{y:5}}
+```
+A downstream operator can then operate on these records,
+for example, merging the two sides of the join using
+spread operators (`...`), i.e.,
+```
+from f1.json
+| cross join (from f2.json) as {f1,f2}
+| values {...f1,...f2}
+```
+produces
+```
+{x:1,y:4}
+{x:2,y:4}
+{x:3,y:4}
+{x:1,y:5}
+{x:2,y:5}
+{x:3,y:5}
+```
+In contrast, relational scoping using identifier scoping in a `SELECT` clause
+with the table source identified in `FROM` and `JOIN` clauses, e.g., this query
+produces the same result:
+```
+SELECT f1.x, f2.y FROM f1.json as f1 CROSS JOIN f2.json as f2
+```
+
+##### Subquery Scoping
+
+A subquery embedded in an expression can also combine data entities
+via pipe scoping as in
+```
+from outer | values {outer:this,inner:[from inner | ...]}
```

Comment:
Would you be game to change the `from` targets to names like `outer_source` or `outer_input` etc.? The way the targets currently are the same as the record keys seems like it has the potential to be confusing.